presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoins.java

```diff
@@ -530,39 +660,111 @@ public boolean equals(Object obj)
+
+                Map<VariableReferenceExpression, RowExpression> interAssignments = new HashMap<>();
+                flattenNode(node, sourceLimit, interAssignments);
+
+                // We resolve the intermediate assignments to only inputs of the flattened join node
+                ImmutableSet<VariableReferenceExpression> inputVariables = sources.stream().flatMap(s -> s.getOutputVariables().stream()).collect(toImmutableSet());
+                this.intermediateAssignments = resolveAssignments(interAssignments, inputVariables);
+                rewriteFilterWithInlinedAssignments(this.intermediateAssignments);
+            }
+
+            private Assignments resolveAssignments(Map<VariableReferenceExpression, RowExpression> assignments, Set<VariableReferenceExpression> availableVariables)
+            {
+                HashSet<VariableReferenceExpression> resolvedVariables = new HashSet<>();
+                for (VariableReferenceExpression variable : assignments.keySet()) {
+                    resolveVariable(variable, resolvedVariables, assignments, availableVariables);
+                }
+
+                return Assignments.builder().putAll(assignments).build();
+            }
+
+            private void resolveVariable(VariableReferenceExpression variable, HashSet<VariableReferenceExpression> resolvedVariables, Map<VariableReferenceExpression,
+                    RowExpression> assignments, Set<VariableReferenceExpression> availableVariables)
+            {
+                RowExpression expression = assignments.get(variable);
+
+                Sets.SetView<VariableReferenceExpression> variablesToResolve = Sets.difference(Sets.difference(extractUnique(expression), availableVariables), resolvedVariables);
+                if (variablesToResolve.isEmpty()) {
+                    resolvedVariables.add(variable);
+                    return;
+                }
+
+                variablesToResolve.forEach(variableToResolve -> resolveVariable(variableToResolve, resolvedVariables, assignments, availableVariables));
+                // Modify the assignments to replace the variables with the resolved expressions
+                assignments.put(variable, replaceExpression(expression, assignments));
+                // Mark the variable as resolved
+                resolvedVariables.add(variable);
             }
 
-            private void flattenNode(PlanNode node, int limit)
+            private void rewriteFilterWithInlinedAssignments(Assignments assignments)
+            {
+                ImmutableList.Builder<RowExpression> modifiedFilters = ImmutableList.builder();
+                filters.forEach(filter -> modifiedFilters.add(replaceExpression(filter, assignments.getMap())));
+                filters = modifiedFilters.build();
+            }
+
+            private void flattenNode(PlanNode node, int limit, Map<VariableReferenceExpression, RowExpression> assignmentsBuilder)
```

Comment:
Do you mean ImmutableMapBuilder instead of Map here?

Reply:
No, since the intermediate assignments Map is read & updated as we traverse it in the `resolveVariable` method, I wanted to use a mutable HashMap